### PR TITLE
Servlets: NewcomerServlet

### PR DESCRIPTION
**Pull request by denisdimitrov**
Index page that contains 3 radio buttons. If a person visits one of the pages for the first time the text will greet him. If he has already visited the page, the text will say: Welcome back! This goes for each page individually, even for the index one.
screenshot from 2017-01-05 10-52-15
screenshot from 2017-01-05 10-52-33
screenshot from 2017-01-05 10-52-44
screenshot from 2017-01-05 10-53-00
screenshot from 2017-01-05 10-53-12
screenshot from 2017-01-05 10-53-21
screenshot from 2017-01-05 10-53-29
screenshot from 2017-01-05 10-53-44

### CODE REVIEW

File: src/main/java/newcomer/HtmlTemplate.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package newcomer;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>.
+ */
+public class HtmlTemplate {
```
Comment (mgenov) [inline src/main/java/newcomer/HtmlTemplate.java]: Nothing shows that it's HTML related.

File: src/main/java/newcomer/servlet/NewcomerServlet.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+public class NewcomerServlet extends HttpServlet {
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        render(request.getRequestURI().substring(1), request, response);
+    }
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        String param = request.getParameter("button");
+        if (param.equals("firstpage")) {
```
Comment (mgenov) [inline src/main/java/newcomer/servlet/NewcomerServlet.java]: If you are going to add a new page you have to change this piece of code which breaks the Open Closed Principle. 

You can fix this by moving this page lookup one level up.

File: src/main/java/newcomer/servlet/NewcomerServlet.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    private void render(String resourceName, HttpServletRequest request, HttpServletResponse response) {
+        try {
+            response.setContentType("text/html");
+            PrintWriter writer = response.getWriter();
+            URL url = NewcomerServlet.class.getClassLoader().getResource(resourceName);
+            String html = Files.toString(new File(url.getPath()), Charset.defaultCharset());
+            HtmlTemplate template = new HtmlTemplate(html);
+            HttpSession session = request.getSession();
+            if ((session.getAttribute("greeting") == null) || (session.getAttribute("greeting") == "Welcome back!")) {
```
Comment (mgenov) [inline src/main/java/newcomer/servlet/NewcomerServlet.java]: getAttribute("visited") = true/false sounds much reasonable then greeting. 

visited gives additional information to the reader then "greeting".

File: src/main/resources/secondpage.html [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
```
Comment (mgenov) [inline src/main/resources/secondpage.html]: I don't see much difference between pages, so maybe 1 page could be re-used. 

```html
${pageName}
<h1>${greeting}</h1>
```

File: src/test/java/newcomer/FakeSession.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import javax.servlet.http.HttpSessionContext;
+import java.util.Enumeration;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>.
+ */
+public class FakeSession implements HttpSession {
+    private Map<String, Object> map = new LinkedHashMap<>();
```
Comment (mgenov) [inline src/test/java/newcomer/FakeSession.java]: map should be attributes is it holds session attributes.

File: src/test/java/newcomer/IndexPageTest.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>.
+ */
+public class IndexPageTest {
```
Comment (mgenov) [inline src/test/java/newcomer/IndexPageTest.java]: Looks same as FirstPageTest, SecondPageTest and ThirdPageTest.

File: src/main/resources/secondpage.html (hunk 4 above)
Comment (denisdimitrov) [inline src/main/resources/secondpage.html]: Everything else is fixed and will be pushed in the next commit, except this. The task said that there have to be 4 pages total, that's why they kind of look the same. Should I just use one, like you asked?

Comment (mgenov) [inline src/main/resources/secondpage.html]: As user will be able to see 4 pages, doesn't mean that you have to duplicate the same code 4 times.

Comment (denisdimitrov) [inline src/main/resources/secondpage.html]: I see. I will fix it.

**Comment by denisdimitrov** [comment]
I applied the changes and removed the duplication in the html files. However I'm not sure if it's working properly.

**Comment by mgenov** [comment]
I've checkout out the PR and this is what I get when I try to open "/" on a running newcomer app:

```
java.lang.NullPointerException
	at newcomer.servlet.NewcomerServlet.render(NewcomerServlet.java:36)
	at newcomer.servlet.NewcomerServlet.doGet(NewcomerServlet.java:23)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:837)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:584)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:224)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1180)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:512)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
	at org.eclipse.jetty.server.Server.handle(Server.java:534)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:320)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95)
	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589)
	at java.lang.Thread.run(Thread.java:745)
java.io.FileNotFoundException: /Users/mgenov/temp/servlets-part2/build/classes/main (Is a directory)
	at java.io.FileInputStream.open0(Native Method)
	at java.io.FileInputStream.open(FileInputStream.java:195)
	at java.io.FileInputStream.<init>(FileInputStream.java:138)
	at com.google.common.io.Files$FileByteSource.openStream(Files.java:126)
	at com.google.common.io.Files$FileByteSource.openStream(Files.java:116)
	at com.google.common.io.ByteSource$AsCharSource.openStream(ByteSource.java:448)
	at com.google.common.io.CharSource.read(CharSource.java:213)
	at com.google.common.io.Files.toString(Files.java:274)
	at newcomer.servlet.NewcomerServlet.render(NewcomerServlet.java:36)
	at newcomer.servlet.NewcomerServlet.doGet(NewcomerServlet.java:23)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:837)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:584)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:224)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1180)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:512)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
	at org.eclipse.jetty.server.Server.handle(Server.java:534)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:320)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95)
	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by mgenov** [comment]
Here is a short list of things that need to be considered: 

 * IndexPageTest is testing NewcomerServlet which doesn't sound reasonable to me
 * resources folder has no packages and all templates are read from the root package (which is and the cause of the error shown above probably)
 * non existing pages could be requested and app will blow up
 * template is not rendered correctly - **"The last page you were on was: ${page}"**
screen shot 2017-01-05 at 3 53 32 pm

**Comment by denisdimitrov** [comment]
Renamed the class to NewcomerServletTest and now only the 2 pages can be accessed.

**Pull request merged**